xdebug_stack.c

```diff
@@ -1288,6 +1291,22 @@ function_stack_entry *xdebug_add_stack_frame(zend_execute_data *zdata, zend_op_a
 					} else {
 						ZVAL_COPY_VALUE(&(tmp->var[tmp->varc].data), ZEND_CALL_VAR_NUM(zdata, zdata->func->op_array.last_var + zdata->func->op_array.T + i - arguments_wanted));
 					}
+					if (Z_TYPE(tmp->var[tmp->varc].data) == IS_OBJECT) {
+						/* Tell garbage collector to NOT gc the objects until the stack is removed
+						 * Otherwise, Zend GC might remove functions parameters (or maybe other variables content)
+						 * and XDebug will be stuck when trying to dump the stack frame
+						 * because some variables in this stackframe will no longer exists (because being GCed) */
+						#if PHP_VERSION_ID >= 70300
+							GC_ADDREF(tmp->var[tmp->varc].data.value.obj);
+						#else
+							GC_REFCOUNT(tmp->var[tmp->varc].data.value.obj)++;
+						#endif
+
+						/* Save the GC lock references so we can remove them once the frame is discarded */
+						tmp->gc_locked_objects = realloc(tmp->gc_locked_objects, tmp->gc_locked_objects_count + 1);
```

Comment:
It's probably not very efficient to do this for each call, especially if this happens a lot. Most other bits of code use not only a `count` (like you do), but also an allocated `size`. See for example: https://github.com/xdebug/xdebug/blob/master/xdebug_branch_info.c#L163-L165

Reply:
Did you miss this comment, or hadn't you gotten to it yet?

Reply:
I'll leave that to the very last (optimization!) ;)

Reply:
Oooooh, now I get why I had "random" segfaults now: I've switched from x86 to x64, and since I had forgotten the `sizeof(zend_object *)`, then the realloc was generating the segfault I got : )

Reply:
for the efficiency, I see no way to improve this without making code unreadable, if you have clues, please say so. I did a quick stress test and ended up with `100000 * [200 nested function calls]` PHP 7.3.3 64bits, without triggering the segfault this PR fixes (=normal execution with no exception thrown):
- 27-28 seconds without xdebug 2.7 (xdebug.default_enabled=0)
- 28-29 seconds with xdebug 2.7 (xdebug.default_enabled=1)
- 29-30 seconds without patched xdebug (xdebug.default_enabled=0)
- 29-31 seconds with patched xdebug (xdebug.default_enabled=1)

So I would say the overhead is acceptable for a debug extension?